universal-application-tool-0.0.1/app/services/question/types/FileUploadQuestionDefinition.java

```diff
@@ -0,0 +1,97 @@
+package services.question.types;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableMap;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.OptionalLong;
+import models.LifecycleStage;
+import services.Path;
+
+public class FileUploadQuestionDefinition extends QuestionDefinition {
+
+  public FileUploadQuestionDefinition(
+      OptionalLong id,
+      long version,
+      String name,
+      Path path,
+      Optional<Long> repeaterId,
+      String description,
+      LifecycleStage lifecycleStage,
+      ImmutableMap<Locale, String> questionText,
+      ImmutableMap<Locale, String> questionHelpText) {
+    super(
+        id,
+        version,
+        name,
+        path,
+        repeaterId,
+        description,
+        lifecycleStage,
+        questionText,
+        questionHelpText,
+        FileUploadValidationPredicates.create());
+  }
+
+  public FileUploadQuestionDefinition(
+      long version,
+      String name,
+      Path path,
+      Optional<Long> repeaterId,
+      String description,
+      LifecycleStage lifecycleStage,
+      ImmutableMap<Locale, String> questionText,
+      ImmutableMap<Locale, String> questionHelpText) {
+    super(
+        version,
+        name,
+        path,
+        repeaterId,
+        description,
+        lifecycleStage,
+        questionText,
+        questionHelpText,
+        FileUploadValidationPredicates.create());
+  }
+
+  @AutoValue
+  public abstract static class FileUploadValidationPredicates extends ValidationPredicates {
+
+    public static FileUploadValidationPredicates parse(String jsonString) {
+      try {
+        return mapper.readValue(
+            jsonString,
+            AutoValue_FileUploadQuestionDefinition_FileUploadValidationPredicates.class);
+      } catch (JsonProcessingException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    public static FileUploadValidationPredicates create() {
+      return new AutoValue_FileUploadQuestionDefinition_FileUploadValidationPredicates();
+    }
+  }
+
+  public FileUploadValidationPredicates getFileUploadValidationPredicates() {
+    return (FileUploadValidationPredicates) getValidationPredicates();
+  }
+
+  @Override
+  public QuestionType getQuestionType() {
+    return QuestionType.FILEUPLOAD;
+  }
+
+  @Override
+  ImmutableMap<Path, ScalarType> getScalarMap() {
+    return ImmutableMap.of(getFileKeyPath(), getFileKeyType());
+  }
+
+  public Path getFileKeyPath() {
```

Comment:
question: what does a file key represent? Is it the name of the file or some kind of clue to where it is located in S3?

Reply:
Yeah, right now I don't have a clear idea what it will be yet, but it will be some unique identifier to reference the file/object in S3, I am thinking something like `<application id>-<formatted filename>`.